median/report_codec.go

```diff
@@ -0,0 +1,50 @@
+package median
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/libocr/offchainreporting2/reportingplugin/median"
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+)
+
+const typeName = "MedianReport"
+
+func newReportCodec(codec types.Codec) (median.ReportCodec, error) {
+	if codec == nil {
+		return nil, errors.New("codec cannot be nil")
+	}
```

Comment:
Why does this need a nil check? I think we can drop that, and actually shedding the whole constructor would clean up the callers. e.g.:
```go
factory.ReportCodec = &reportCodec{codec}
```

Reply:
I thought it's best to trust but verify because if it's nil then we would panic later.  My understanding of go conversion was that your struct should work if it's zeroed, or you should provide a constructor [1].  I'll change it if you still want me to though.


[1]
https://go.dev/doc/effective_go#composite_literals

> Constructors and composite literals
> Sometimes the zero value isn't good enough and an initializing constructor is necessary, as in this example derived from package os.

```
> func NewFile(fd int, name string) *File {
>     if fd < 0 {
>         return nil
>     }
>     f := File{fd, name, nil, 0}
>     return &f
> }
```

Reply:
I don't think it follows that you must create a constructor when the zero value is not immediately useful, but even if we accept that, I don't think it is correct to say the zero value is not useful, because we have access to the unexported API and are free to set the field during or after initialization.

Separately, it is not clear to me when this nil checking rule would be applied. I'm concerned that defaulting to constructors and nil checks everywhere would snowball into Java verbosity.